waveformview/src/main/java/space/siy/waveformview/FixedWaveFormView.kt

```diff
@@ -0,0 +1,266 @@
+          for (i in 0 until possibleBlockCountOnScreen) {
+            resampleData[i] = value.samples.average(
+                i * numberOfDataToNormalize, (i + 1) * numberOfDataToNormalize
+            )
+          }
+        }
+      }
+    }
+
+  /**
+   * position in milliseconds
+   */
+  var position: Long = 0
+    set(value) {
+      if (seeking) return
+      field = value
+      if (position == 0L) {
+        lastProgress = 0
+        seekingPosition = 0
+      } else if (position == data?.duration) {
+        seekingPosition = position
+      }
+      lastProgress =
+        if (position - seekingPosition > 0) position - seekingPosition else lastProgress
+      seekingPosition += lastProgress
+      invalidate()
+    }
+
+  private var lastProgress = 0L
+
+  /**
+   * Width each block
+   */
+  var blockWidth = lp.getDimension(R.styleable.FixedWaveFormView_blockWidth, 5f)
+    set(value) {
+      field = value
+      blockPaint.strokeWidth = blockWidth - SPLIT_GAP
+    }
+  /**
+   * @see Callback
+   */
+  var callback: Callback? = null
+
+  /**
+   * Scale of top blocks
+   */
+  var topBlockScale = lp.getFloat(R.styleable.FixedWaveFormView_topBlockScale, 1f)
+  /**
+   * Scale of bottom blocks
+   */
+  var bottomBlockScale = lp.getFloat(R.styleable.FixedWaveFormView_bottomBlockScale, 0f)
+
+  /**
+   * Color used in played blocks
+   */
+  var blockColorPlayed: Int = lp.getColor(R.styleable.FixedWaveFormView_blockColorPlayed, Color.RED)
+    set(value) {
+      field = value
+      barShader = LinearGradient(canvasWidth / 2f - 1, 0f, canvasWidth / 2f + 1, 0f, blockColorPlayed, blockColor, Shader.TileMode.CLAMP)
+      blockPaint.shader = barShader
+    }
+
+  /**
+   * Color used in blocks default
+   */
+  var blockColor: Int = lp.getColor(R.styleable.FixedWaveFormView_blockColor, Color.WHITE)
+    set(value) {
+      field = value
+      barShader = LinearGradient(canvasWidth / 2f - 1, 0f, canvasWidth / 2f + 1, 0f, blockColorPlayed, blockColor, Shader.TileMode.CLAMP)
+      blockPaint.shader = barShader
+    }
+
+  /**
+   * The resampled data to show
+   *
+   * This generate when [data] set
+   */
+  private var resampleData = FloatArray(0)
+
+  private val blockPaint = Paint()
+
+  private var offsetX = 0f
+  private var canvasWidth = 0
+  private var seekingPosition = 0L
+  private var donePlaying = false
+
+  private var barShader = LinearGradient(0f, 0f, 0f, 700f, Color.RED, Color.GRAY, Shader.TileMode.CLAMP)
+
+  init {
+    blockPaint.strokeWidth = blockWidth - 2
+    blockPaint.shader = barShader
+  }
+
+  @SuppressLint("DrawAllocation")
+  override fun onDraw(canvas: Canvas) {
+    super.onDraw(canvas)
+
+    offsetX = (width / (data?.duration ?: 1L).toFloat()) * seekingPosition
+    // Right now, I don't have any better way than allocating shader in every invalidate() invocation
+    barShader = LinearGradient(offsetX, 0f, offsetX + 1, 0f, blockColorPlayed, blockColor, Shader.TileMode.CLAMP)
```

Comment:
Is it possible to use `matrix` to modify LinearGradient rather than creating new object every time?

Reply:
Does not have any better solution right now, plus it takes < 1ms to do this operation and whole view drawing is way under 16ms in older Nexus 5x devices. View is right now optimized, MediaPlayer is hogging the memory.